src/core/DataFile.cpp

```diff
@@ -328,6 +405,164 @@ bool DataFile::writeFile( const QString& filename )
+	{
+		QDomNodeList list = elementsByTagName(it->first);
+
+		// Go through all elements with the tagname from our map
+		for (int i = 0; !list.item(i).isNull(); ++i)
+		{
+			QDomElement el = list.item(i).toElement();
+
+			std::vector<QString>::const_iterator res = it->second.begin();
+
+			// Search for attributes that point to resources
+			while (res != it->second.end())
+			{
+				// If the element has that attribute
+				if (el.hasAttribute(*res))
+				{
+					// Get absolute path to resource
+					QString resPath = PathUtil::toAbsolute(el.attribute(*res));
+					// If we are running without the project loaded (from CLI), "local:" base
+					// prefixes aren't converted, so we need to convert it ourselves
+					if (PathUtil::baseLookup(resPath) == PathUtil::Base::LocalDir)
+					{
+						resPath = QFileInfo(m_fileName).path() + "/" + resPath.remove(0,
+							PathUtil::basePrefix(PathUtil::Base::LocalDir).length());
+					}
+
+					// Check if we need to add a counter to the filename
+					QString finalFileName = QFileInfo(resPath).fileName();
+					QString extension = resPath.section('.', -1);
+					int repeatedNames = 0;
+					for (QString name : namesList)
+					{
+						if (finalFileName == name)
+						{
+							++repeatedNames;
+						}
+					}
+					// Add the name to the list before modifying it
+					namesList.push_back(finalFileName);
+					if (repeatedNames)
+					{
+						// Remove the extension, add the counter and add the
+						// extension again to get the final file name
+						finalFileName.truncate(finalFileName.lastIndexOf('.'));
+						finalFileName = finalFileName + "-" + QString::number(repeatedNames) + "." + extension;
+					}
+
+					// Final path is our resources dir + the new file name
+					QString finalPath = resourcesDir + "/" + finalFileName;
+
+					// Copy resource file to the resources folder
+					if(!QFile::copy(resPath, finalPath))
+					{
+						qWarning("ERROR: Failed to copy resource");
+						return false;
+					}
+
+					// Update attribute path to point to the bundle file
+					QString newAtt = PathUtil::basePrefix(PathUtil::Base::LocalDir) + "resources/" + finalFileName;
+					el.setAttribute(*res, newAtt);
+				}
+				++res;
+			}
+		}
+		++it;
+	}
+
+	return true;
+}
+
+
+
+
+/* @brief This recursive method will go through all XML nodes of the DataFile
+ *        and check whether any of them have local paths. If they are not on
+ *        our list of elements that can have local paths we return true,
+ *        indicating that we potentially have plugins with local paths that
+ *        would be a security issue. The Song class can then abort loading
+ *        this project.
+ * @param QDomElement parent The parent node being interated. When called
+ *        without arguments, this will be an empty element that will be
+ *        ignored (since the second parameter will be true).
+ * @param bool firstCall Defaults to true, and indicates to this recursive
```

Comment:
Sorry, I forgot to fix the default parameters comment on the CPP file. The default `parent` value is `QDomElement()` (an empty element), not `documentElement()` (because we can't call it without an object, hence can't use it as a default parameter). I used `firstCall` because there's no way to differentiate between the default empty element and an empty element from the XML project file (this was what caused the infinite recursivity I posted on Discord), so I used this little variable to discern between the method being called for the first time and it being called from one of the recursive iterations.

Reply:
Understood - you recursed infinitely because everytime you hit an empty `QDomElement`, your code did reset the node to the `documentElement`. Resolving.